File: api/app/schemas/ticket.py
from __future__ import annotations

from datetime import UTC, datetime

from pydantic import BaseModel, computed_field

_TERMINAL_STATUSES = {"pending_closure", "resolved", "closed", "cancelled"}
_PAUSED_STATUSES = {"waiting_client"}


class TicketEventResponse(BaseModel):
    id: int
    event_type: str
    payload: dict | None
    occurred_at: datetime

    model_config = {"from_attributes": True}


class AssigneeResponse(BaseModel):
    id: int
    name: str

    model_config = {"from_attributes": True}


class TicketResponse(BaseModel):
    id: int
    source_id: int
    source_name: str
    external_id: str
    type: str | None
    priority: str | None
    status: str
    subject: str
    description: str | None
    source_metadata: dict | None
    source_created_at: datetime | None
    source_updated_at: datetime | None
    first_ingested_at: datetime
    last_synced_at: datetime
    sla_due_at: datetime | None = None
    sla_started_at: datetime | None = None
    sla_paused_seconds: int = 0
    sla_paused_since: datetime | None = None
    last_inbound_at: datetime | None = None
    assigned_to: AssigneeResponse | None = None

    @computed_field  # type: ignore[prop-decorator]
    @property
    def sla_status(self) -> str | None:
        if self.sla_due_at is None:
            return None
        if self.status in _TERMINAL_STATUSES:
            return "met"
        if self.status in _PAUSED_STATUSES and self.sla_paused_since is not None:
            return "paused"
        now = datetime.now(UTC)
        if now >= self.sla_due_at:
            return "overdue"
        # Use sla_started_at for the total span if available, else fall back to ingestion
        reference = self.sla_started_at or self.first_ingested_at
        total = (self.sla_due_at - reference).total_seconds()
        remaining = (self.sla_due_at - now).total_seconds()
        if total > 0 and remaining / total < 0.2:
            return "at_risk"
        return "on_time"

    model_config = {"from_attributes": True}


class AssignTicketRequest(BaseModel):
    user_id: int | None


class UpdateStatusRequest(BaseModel):
    status: str
    comment: str | None = None


class OverrideSlaRequest(BaseModel):
    due_at: datetime
    reason: str | None = None


class TicketDetailResponse(TicketResponse):
    events: list[TicketEventResponse]


class TicketListResponse(BaseModel):
    items: list[TicketResponse]
    total: int
    limit: int
    offset: int

File: frontend/src/types/ticket.ts
export type TicketType = 'BUG' | 'IMPROVEMENT' | 'QUESTION' | 'SUPPORT'
export type TicketPriority = 'URGENT' | 'HIGH' | 'MEDIUM' | 'LOW'
export type TicketStatus =
  | 'OPEN'
  | 'IN_PROGRESS'
  | 'WAITING_DEV'
  | 'IN_DEV'
  | 'PENDING_CLOSURE'
  | 'RESOLVED'
  | 'CLOSED'
  | 'CANCELLED'
  | 'MERGED'

export interface TicketEvent {
  id: number
  event_type: string
  payload: Record<string, unknown> | null
  occurred_at: string
}

export interface Assignee {
  id: number
  name: string
}

export interface Tag {
  id: number
  name: string
  color: string
  description: string | null
}

export interface ChecklistItem {
  id: number
  text: string
  is_done: boolean
  position: number
  created_by: number | null
  done_by: number | null
  done_at: string | null
  created_at: string
}

export interface ChecklistProgress {
  done: number
  total: number
}

export interface Ticket {
  id: number
  source_id: number
  source_name: string
  external_id: string
  type: string | null
  priority: string | null
  status: string
  subject: string
  description: string | null
  source_metadata: Record<string, unknown> | null
  source_created_at: string | null
  source_updated_at: string | null
  first_ingested_at: string
  last_synced_at: string
  sla_due_at: string | null
  sla_started_at: string | null
  sla_paused_since: string | null
  sla_status: 'on_time' | 'at_risk' | 'overdue' | 'met' | 'paused' | null
  last_inbound_at: string | null
  assigned_to: Assignee | null
  tags: Tag[]
  merged_into_ticket_id: number | null
  merged_at: string | null
  deployment_scheduled_at: string | null
  pr_number: string | null
  checklist_progress: ChecklistProgress | null
}

export interface TicketDetail extends Ticket {
  events: TicketEvent[]
  checklist_items: ChecklistItem[]
}

export interface TicketListResponse {
  items: Ticket[]
  total: number
  limit: number
  offset: number
}

export interface TicketMessage {
  id: number
  direction: 'inbound' | 'outbound'
  author_name: string
  author_user_id: number | null
  body: string
  is_internal: boolean
  mentioned_user_ids: number[]
  created_at: string
  attachments?: {
    id: number
    filename: string
    content_type: string
    size_bytes: number
    download_url: string
  }[]
}

export interface TicketAttachment {
  id: number
  ticket_id: number
  original_filename: string
  content_type: string
  size_bytes: number
  created_at: string
  download_url: string
}

export interface NoteAuthor {
  id: number
  name: string
}

export interface Note {
  id: number
  ticket_id: number
  body: string
  author: NoteAuthor | null
  created_at: string
}

export interface TicketFilters {
  source_id?: number
  status?: string
  active_only?: boolean
  priority?: string
  type?: string
  search?: string
  assigned_to_user_id?: number
  unassigned?: boolean
  tag_ids?: number[]
  ticket_ids?: number[]
  limit?: number
  offset?: number
}
